File: leetcodes/94. Binary Tree Inorder Traversal/BinaryTreeInorderTraversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    #recursion  timeO(n) , space O(n) -> singled side tree is O(n), average tree is O(logN)
    def __init__(self):
        #create List[int] res as new array, use helper(root, res) to traverse result into res, return res
        self.traversal = []
    #helper(root,res)
    def dfs(self, node):
    #if root exists, 
        if node.left:
        #if root.left exists, helper(root.left, res)
            self.dfs(node.left)
        #add root.val into res
        self.traversal.append(node.val)
        #if root.right exists, helper(root.right, res)
        if node.right:
            self.dfs(node.right)
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        #input: tree node, output: list
        if root:
            self.dfs(root)
        return self.traversal

#iteration: time O(n) , space O(n)
class Solution:
    def __init__(self):
        self.traversal = []
    
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        stack = []
        curr = root
        while curr or stack:
            while curr:
                # find the left-most node, while storing root nodes in stack
                stack.append(curr)
                curr = curr.left
            curr = stack.pop()
            self.traversal.append(curr.val)
            curr = curr.right
            
        return self.traversal

File: leetcodes/94. Binary Tree Inorder Traversal/binaryTreeInorderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
 var inorderTraversal = function(root) {
  //input: treenode
  //output: int[]
  /**
   * order: left->root->right
   * 1)recursive timeO(n), space O(logn) ->O(n)
   *
   * 2)iterating method using stack
   *
   * 3)morris traversal
   */
  let res = [];
  helper(root, res);
  return res;
};
const helper = (root, res) =>{
  if(root) {
    if (root.left) helper(root.left,res);
    res.push(root.val);
    if(root.right) helper(root.right,res);
  }
}